File: source/utilities.lua
local String = require(script.Parent.string)

return { 
	stringEncaseCRLFWithFirstIndex = function(str, prefix, postfix, index)
		local endIndex = 1
		local returnValue = ""
		repeat
			local gotCR = string.sub(str, index - 1, index - 1) == "\r"
			returnValue ..= String.slice(str, endIndex, if gotCR then index - 2 else index - 1) .. prefix .. (if gotCR
			then "\r\n"
			else "\n") .. postfix
		endIndex = index + 1
		-- TODO: add String.indexOf and use it here
		index = string.find(string_, "\n", endIndex)
		until index == nil

	returnValue += String.slice(str, endIndex)
	return returnValue
end
}

File: source/utilities.js
// TODO: When targeting Node.js 16, use `String.prototype.replaceAll`.
export function stringReplaceAll(string, substring, replacer) {
	const index = string.indexOf(substring);
	if (index === -1) {
		return string;
	}

	const substringLength = substring.length;
	const replacement = substring + replacer;
	let result = '';
	let lastIndex = 0;
	let currentIndex = index;

	do {
		result += string.slice(lastIndex, currentIndex) + replacement;
		lastIndex = currentIndex + substringLength;
		currentIndex = string.indexOf(substring, lastIndex);
	} while (currentIndex !== -1);

	result += string.slice(lastIndex);
	return result;
}

export function stringEncaseCRLFWithFirstIndex(string, prefix, postfix, index) {
	let endIndex = 0;
	let returnValue = '';
	do {
		const gotCR = string[index - 1] === '\r';
		returnValue += string.slice(endIndex, (gotCR ? index - 1 : index)) + prefix + (gotCR ? '\r\n' : '\n') + postfix;
		endIndex = index + 1;
		index = string.indexOf('\n', endIndex);
	} while (index !== -1);

	returnValue += string.slice(endIndex);
	return returnValue;
}
